src/langchain_google_memorystore_redis/chat_message_history.py

```diff
@@ -0,0 +1,70 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import redis
+from typing import List, Optional
+
+from langchain_core.chat_history import BaseChatMessageHistory
+from langchain_core.messages import (
+    BaseMessage,
+    message_to_dict,
+    messages_from_dict,
+)
+
+
+class MemorystoreChatMessageHistory(BaseChatMessageHistory):
+    """Chat message history stored in a Cloud Memorystore for Redis database."""
+
+    def __init__(
+        self,
+        client: redis.Redis,
+        session_id: str,
+        ttl: Optional[int] = None,
+    ):
+        """Initializes the chat message history for Memorystore for Redis.
+
+        Args:
+            client: A redis.Redis client object.
+            session_id: A string that uniquely identifies the chat history.
+            ttl: Specifies the time in seconds after which the session will
+                expire and be eliminated from the Redis instance since the most
+                recent message is added.
+        """
+
+        self.redis = client
+        self.key = session_id
+        self.ttl = ttl
+
+    def __del__(self):
+        self.redis.close()
+
+    @property
+    def messages(self) -> List[BaseMessage]:
+        """Retrieve all messages chronologically stored in this session."""
+        all_elements = self.redis.lrange(self.key, 0, -1)
+        messages = messages_from_dict(
+            [json.loads(e.decode("utf-8")) for e in all_elements]
```

Comment:
any reason to pass in "utf-8" specifically?

Reply:
Redis stores all keys and values in bytes - it encodes every string to bytes with utf-8 so we need to decode it to a native string in Python for input/output

Reply:
Are we encoding it or is the redis client encoding it (I don't see encode called below)?

Does the redis client explictly call out it encodes in "utf-8"? Or does it use the default for python (which happens to be utf-8 atm)?  https://docs.python.org/3/library/stdtypes.html#bytes.decode

If they are using the default, we should too

Reply:
Great catch. utf-8 is the [default encoding of the Python redis-py package](https://redis-py.readthedocs.io/en/stable/connections.html#generic-client). So we should use the same encoder as the user uses. Updated.